src/components/include/cbuf.h

```diff
@@ -262,63 +252,38 @@ cbuf_vect_lookup_addr(long idx, int tmem)
  * returns an error (NULL) if called by the owner of the cbuf_t.
  */
 static inline void * 
-__cbuf2buf(cbuf_t cb, int len, int tmem)
+cbuf2buf(cbuf_t cb, int len)
 {
 	u32_t id;
 	struct cbuf_meta *cm;
-	union cbufm_info ci;//, ci_new;
 	void *ret = NULL;
 	long cbidx;
 	int t;
 
 	if (unlikely(!len)) return NULL;
-	cbuf_unpack(cb, &id, &t);
-	assert(t == tmem);
-
-	CBUF_TAKE();
+	cbuf_unpack(cb, &id);
 	cbidx = cbid_to_meta_idx(id);
 again:
 	do {
-		cm = cbuf_vect_lookup_addr(cbidx, tmem);
-		if (unlikely(!cm || cm->nfo.v == 0)) {
-			if (__cbuf_2buf_miss(id, len, tmem)) goto done;
+		cm = cbuf_vect_lookup_addr(cbidx);
+		if (unlikely(!cm || cm->nfo == 0)) {
+			if (__cbuf_2buf_miss(id, len)) goto done;
 			goto again;
 		}
-	} while (unlikely(!cm->nfo.v));
-	ci.v = cm->nfo.v;
+	} while (unlikely(!cm->nfo));
 	/* shouldn't cbuf2buf your own buffer! */
-	if (unlikely(CBUF_OWNER(cm->nfo.c.flags))) goto done;
-
-	if (!tmem) {
-		if (unlikely(cm->nfo.c.flags & CBUFM_TMEM)) goto done;
-		if (unlikely((len >> PAGE_ORDER) > cm->sz)) goto done;
-		if(cm->nfo.c.refcnt == CBUFP_REFCNT_MAX)
-			assert(0);
-		cm->nfo.c.refcnt++;
-		assert(cm->owner_nfo.c.nrecvd < TMEM_SENDRECV_MAX);
-		cm->owner_nfo.c.nrecvd++;
-	} else {
-		if (unlikely(!(cm->nfo.c.flags & CBUFM_TMEM))) goto done;
-		if (unlikely(len > PAGE_SIZE)) goto done;
-	}
-	/* if (unlikely(cm->sz && (((int)cm->sz)<<PAGE_ORDER) < len)) goto done; */
-	/* ci_new.v       = ci.v; */
-	/* ci_new.c.flags = ci.c.flags | CBUFM_RECVED | CBUFM_IN_USE; */
-	/* if (unlikely(!cos_cas((unsigned long *)&cm->nfo.v,  */
-	/* 		      (unsigned long)   ci.v,  */
-	/* 		      (unsigned long)   ci_new.v))) goto again; */
-	ret = ((void*)(cm->nfo.c.ptr << PAGE_ORDER));
+	assert(cm->cbid.cbid == id);
+	if (unlikely(CBUF_OWNER(cm))) goto done;
+	if (unlikely((len >> PAGE_ORDER) > cm->sz)) goto done;
+	assert(CBUFM_GET_REFCNT(cm) < CBUF_REFCNT_MAX);
+	CBUFM_INC_REFCNT(cm);
+	assert(cm->owner_nfo.c.nrecvd < CBUF_SENDRECV_MAX);
+	cm->owner_nfo.c.nrecvd++;      //???
```

Comment:
If you have a question or comment, please actually write it down.  This `//???` is cryptic.